### ISSUE TITLE
asking for NVDIMM cache layer enablement

### ISSUE THREAD

**Issue by gaowayne** [apache/trafficserver#8693]
hello dear product owners,
can I ask if this ATS server can enable NVDIMM as cache, it can works with pmem fsdax mode naturally.

**Comment by bryancall** [comment]
We support cache as a block device or a file.  Depends on the OS and how it handles the NVDIMM.

**Comment by gaowayne** [comment]
@bryancall I see ATS support file or block device, I am sorry I am new to this, I did not see any tier at the block cache level. 1st layer is ram_cache, 2nd layer is block or file, if we put nvdimm or pmem into 2nd layer, they are small and expensive, better more hot data is in nvdimm or pmem, more cold data into HDD or QLC

**Comment by cukiernik** [comment]
@gaowayne, I tested https://github.com/apache/trafficserver/pull/8534 using NVDIMM, achieving a lag reduction.

**Comment by bryancall** [comment]
@gaowayne Yes, we don't have a way to tier the block storage for ATS.  You can partition your cache based on hostnames if that help and store certain domains on the NVDIMM block devices.
https://docs.trafficserver.apache.org/admin-guide/storage/index.en.html#partitioning-the-cache

**Comment by cukiernik** [comment]
@gaowayne .Can you verify and share a benchmark to https://github.com/cukiernik/trafficserver/tree/devdax?

**Comment by gaowayne** [comment]
> @gaowayne .Can you verify and share a benchmark to https://github.com/cukiernik/trafficserver/tree/devdax?
buddy, could you please share me quick starter benchmark steps? we can test this and share result here.

**Comment by cukiernik** [comment]
For monitoring IPM You can use: 
`watch tools/ipm-performance`
Simple test:
```
seq 1 1000000 | while read N ; do
 echo "http://127.0.0.1:8080/${RANDOM}"
done | time xargs -P100 -n1 wget -O /dev/null --no-check-certificate 
```

**Comment by cukiernik** [comment]
Hi @gaowayne 
I completed a [autoreconf and rules to initialize devices](https://github.com/cukiernik/trafficserver/tree/devdax), but it still waiting for https://github.com/apache/trafficserver/pull/8534

**Comment by gaowayne** [comment]
> @gaowayne .Can you verify and share a benchmark to https://github.com/cukiernik/trafficserver/tree/devdax?

sorry, we have give up Optane. I am so sorry for late response, I just read this. is it possible ATS do some tuning that leverage QLC SSD well. its seq write is 3.5GB/s random read is 6GB/s. if we do some improvement on the SW together. I can help do benchmark and we can release white paper together?  we have slack to talk detail?

**Comment by bryancall** [comment]
Apache Traffic Server still has no built-in cache storage tiering: there is no mechanism in the cache to keep hot objects on fast media (NVDIMM or persistent memory) and spill cold objects to HDD or QLC. The closest available approach remains partitioning the cache by hostname so specific domains can be assigned to specific (faster) storage, as documented at https://docs.trafficserver.apache.org/en/latest/admin-guide/storage/index.en.html. The community experiment that explored this (the devdax branch) depended on https://github.com/apache/trafficserver/pull/8534, which has since been closed without merging, and the original requester noted they moved off Optane and shifted to a more open-ended QLC SSD tuning idea that never produced a concrete proposal or pull request. With no actionable scope, no associated change in progress, and the issue auto-marked stale twice, I am closing this. If there is interest in real cache tiering with a concrete design, please open a fresh issue (or a design proposal) describing the intended tiering policy, and we can scope it from there.
